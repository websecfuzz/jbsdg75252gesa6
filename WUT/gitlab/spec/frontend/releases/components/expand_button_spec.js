import { mount } from '@vue/test-utils';
import { nextTick } from 'vue';
import ExpandButton from '~/releases/components/expand_button.vue';

const text = {
  expanded: 'Expanded!',
  short: 'Short',
};

describe('Expand button', () => {
  let wrapper;

  const expanderPrependEl = () => wrapper.find('.js-text-expander-prepend');
  const expanderAppendEl = () => wrapper.find('.js-text-expander-append');

  const factory = (options = {}) => {
    wrapper = mount(ExpandButton, {
      ...options,
    });
  };

  beforeEach(() => {
    factory({
      slots: {
        expanded: `<p>${text.expanded}</p>`,
      },
    });
  });

  it('renders the prepended collapse button', () => {
    expect(expanderPrependEl().isVisible()).toBe(true);
    expect(expanderAppendEl().isVisible()).toBe(false);
  });

  it('renders no text when short text is not provided', () => {
    expect(wrapper.findComponent(ExpandButton).text()).toBe('');
  });

  it('does not render expanded text', () => {
    expect(wrapper.findComponent(ExpandButton).text().trim()).not.toBe(text.short);
  });

  describe('when short text is provided', () => {
    beforeEach(() => {
      factory({
        slots: {
          expanded: `<p>${text.expanded}</p>`,
          short: `<p>${text.short}</p>`,
        },
      });
    });

    it('renders short text', () => {
      expect(wrapper.findComponent(ExpandButton).text().trim()).toBe(text.short);
    });

    it('renders button before text', () => {
      expect(expanderPrependEl().isVisible()).toBe(true);
      expect(expanderAppendEl().isVisible()).toBe(false);
      expect(wrapper.findComponent(ExpandButton).element).toMatchSnapshot();
    });
  });

  describe('on click', () => {
    beforeEach(async () => {
      expanderPrependEl().trigger('click');
      await nextTick();
    });

    afterEach(() => {
      expanderAppendEl().trigger('click');
    });

    it('renders only the append collapse button', () => {
      expect(expanderAppendEl().isVisible()).toBe(true);
      expect(expanderPrependEl().isVisible()).toBe(false);
    });

    it('renders the expanded text', () => {
      expect(wrapper.findComponent(ExpandButton).text()).toContain(text.expanded);
    });

    describe('when short text is provided', () => {
      beforeEach(async () => {
        factory({
          slots: {
            expanded: `<p>${text.expanded}</p>`,
            short: `<p>${text.short}</p>`,
          },
        });

        expanderPrependEl().trigger('click');
        await nextTick();
      });

      it('only renders expanded text', () => {
        expect(wrapper.findComponent(ExpandButton).text().trim()).toBe(text.expanded);
      });

      it('renders button after text', () => {
        expect(expanderPrependEl().isVisible()).toBe(false);
        expect(expanderAppendEl().isVisible()).toBe(true);
        expect(wrapper.findComponent(ExpandButton).element).toMatchSnapshot();
      });
    });
  });

  describe('append button', () => {
    beforeEach(async () => {
      expanderPrependEl().trigger('click');
      await nextTick();
    });

    it('clicking hides itself and shows prepend', async () => {
      expect(expanderAppendEl().isVisible()).toBe(true);
      expanderAppendEl().trigger('click');

      await nextTick();
      expect(expanderPrependEl().isVisible()).toBe(true);
    });

    it('clicking hides expanded text', async () => {
      expect(wrapper.findComponent(ExpandButton).text().trim()).toBe(text.expanded);
      expanderAppendEl().trigger('click');

      await nextTick();
      expect(wrapper.findComponent(ExpandButton).text().trim()).not.toBe(text.expanded);
    });

    describe('when short text is provided', () => {
      beforeEach(async () => {
        factory({
          slots: {
            expanded: `<p>${text.expanded}</p>`,
            short: `<p>${text.short}</p>`,
          },
        });

        expanderPrependEl().trigger('click');
        await nextTick();
      });

      it('clicking reveals short text', async () => {
        expect(wrapper.findComponent(ExpandButton).text().trim()).toBe(text.expanded);
        expanderAppendEl().trigger('click');

        await nextTick();
        expect(wrapper.findComponent(ExpandButton).text().trim()).toBe(text.short);
      });
    });
  });
});
