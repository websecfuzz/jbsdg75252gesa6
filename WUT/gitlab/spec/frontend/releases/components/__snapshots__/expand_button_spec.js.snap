// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Expand button on click when short text is provided renders button after text 1`] = `
<span>
  <button
    aria-label="Click to expand text"
    class="btn btn-default btn-default-tertiary btn-icon btn-md button-ellipsis-horizontal gl-button js-text-expander-prepend"
    style="display: none;"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="gl-button-icon gl-fill-current gl-icon s16"
      data-testid="ellipsis_h-icon"
      role="img"
    >
      <use
        href="file-mock#ellipsis_h"
      />
    </svg>
  </button>
  <span>
    <p>
      Expanded!
    </p>
  </span>
  <button
    aria-label="Click to expand text"
    class="btn btn-default btn-default-tertiary btn-icon btn-md button-ellipsis-horizontal gl-button js-text-expander-append"
    style=""
    type="button"
  >
    <svg
      aria-hidden="true"
      class="gl-button-icon gl-fill-current gl-icon s16"
      data-testid="ellipsis_h-icon"
      role="img"
    >
      <use
        href="file-mock#ellipsis_h"
      />
    </svg>
  </button>
</span>
`;

exports[`Expand button when short text is provided renders button before text 1`] = `
<span>
  <button
    aria-label="Click to expand text"
    class="btn btn-default btn-default-tertiary btn-icon btn-md button-ellipsis-horizontal gl-button js-text-expander-prepend"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="gl-button-icon gl-fill-current gl-icon s16"
      data-testid="ellipsis_h-icon"
      role="img"
    >
      <use
        href="file-mock#ellipsis_h"
      />
    </svg>
  </button>
  <span>
    <p>
      Short
    </p>
  </span>
  <button
    aria-label="Click to expand text"
    class="btn btn-default btn-default-tertiary btn-icon btn-md button-ellipsis-horizontal gl-button js-text-expander-append"
    style="display: none;"
    type="button"
  >
    <svg
      aria-hidden="true"
      class="gl-button-icon gl-fill-current gl-icon s16"
      data-testid="ellipsis_h-icon"
      role="img"
    >
      <use
        href="file-mock#ellipsis_h"
      />
    </svg>
  </button>
</span>
`;
